### ISSUE TITLE
Make jQuery.when always async

### ISSUE THREAD

**Issue by markelog** [jquery/jquery#3102]
See discussion in https://github.com/jquery/jquery/issues/3100.

Making it a blocker at the moment for 3.0, can provide a pull for that. 

If there strong motivation to not deal with it, feel free to close.

**Comment by marcandre** [comment]
> It's important to realize that jQuery.when is polymorphic—multi-argument calls being analogous to ES6 `Promise.all` and other calls being analogous to ES6 `Promise.resolve`.

Thanks for the replies.

I'd like to point out that `$.when(one_argument)` is *not* really analogous to `Promise.resolve`, in that it currently returns a wrapped promise while `Promise.resolve` doesn't wrap existing promises:

```
p = Promise.resolve(42)
p === Promise.resolve(p) // => true
j = $.when(42)
j === $.when(j) // => false
```

The argument about the "reliance on `.then` to unwrap secondary thenables" seems to apply to `Promise.all` at least as much when it has more than one argument than when it has a single one.

What I mean to say is that if I had gotten ABDC in the above example (instead of ACDB), then I could understand the reasoning (and disagree with the decision), but currently the justification and behavior seems reversed!

ABCD: Synchronous behavior (what I'm secretly hoping for)
ADBC: Asynchronous behavior (I would secretly disapprove)
ABDC: Inconsistent behavior of single argument vs multiple, with single promise treated specially like `Promise.resolve` (I would disapprove)
ACDB: Inconsistent behavior where single promise is treated differently and illogically (current behavior to which I can only object)

I think there's clearly a bug currently:

    $.when(42).state() // => 'pending', should be 'resolved'

In closing, I'd ask naively: why not have two different methods? `$.when` could be `Promise.all`, while `$.resolve` would be `Promise.resolve`.
`resolve` would coerce its sole argument into a promise.
`when` coerces them via `resolve` if need be. It could keep the synchronous behavior of `done` intact when the given promises have `state` method available.

**Comment by markelog** [comment]
> $.when(42).state()

This would execute asynchronously, so it seems `state` returns correct result. 

> ADBC: Asynchronous behavior (I would secretly disapprove)

That would be with the spirit of the spec and in line with other implementations. 

> why not have two different methods?

I guess we could consider adding new method, but definitely not for 3.0

**Comment by gibson042** [comment]
> I'd like to point out that `$.when(one_argument)` is *not* really analogous to `Promise.resolve`, in that it currently returns a wrapped promise while `Promise.resolve` doesn't wrap existing promises:

```
p = Promise.resolve(42)
p === Promise.resolve(p) // => true
j = $.when(42)
j === $.when(j) // => false
```

That was fixed by 356a3bccb0e7468a2c8ce7d8c9c6cd0c5d436b8b: https://jsfiddle.net/q0fmmv8b/ . 

> The argument about the "reliance on `.then` to unwrap secondary thenables" seems to apply to `Promise.all` at least as much when it has more than one argument than when it has a single one.

My focus was on fixing the already-broken single-argument case, but that is certainly a valid point. I didn't implement it because I didn't want to totally destroy access to the secondary thenables (the `.done` trick we use for an analogous case with `jQuery.Deferred` doesn't work when the resolution value is an array).

> In closing, I'd ask naively: why not have two different methods? `$.when` could be `Promise.all`, while `$.resolve` would be `Promise.resolve`.

I intend to do just that, but in 3.1 or later.

> `when` coerces them via `resolve` if need be. It could keep the synchronous behavior of `done` intact when the given promises have `state` method available.

I don't see that happening, because promises can resolve to thenables.

**Comment by marcandre** [comment]
>> $.when(42).state()
> This would execute asynchronously, so it seems state returns correct result.

I'm not sure I follow. I would expect that the result should be the same as:
```
d = $.Deferred()
d.resolveWith(42)
p = d.promise()
p.state() // => 'resolved'
```

Mozilla's doc states that "Promise.resolve(value) method returns a Promise object that is resolved with the given value. If the value is a thenable [...]; otherwise the returned promise will be fulfilled with the value."

I would expect the equivalent to have the corresponding state 'resolved', not 'pending'.

>> ADBC: Asynchronous behavior (I would secretly disapprove)
> That would be with the spirit of the spec and in line with other implementations.

The spec and other implementations worry about `then` handlers being run asynchronously, not about the state of the promise itself. It's only with `jQuery`'s `done` and `state` methods that we can talk about synchronous behavior. Using `then` on jQuery would still be asynch and in line with other implementations even in the ABCD case.

**Comment by marcandre** [comment]
> That was fixed by 356a3bc: https://jsfiddle.net/q0fmmv8b/ .

I'm not sure I made myself clear, but `Promise.resolve(somePromise)` returns the *exact same object* `somePromise`, not a new promise. `$.when(aPromise)` returns a *different* object, that doesn't even have the same `state`.

>> It could keep the synchronous behavior of done intact when the given promises have state method available.
> I don't see that happening, because promises can resolve to thenables.

Indeed, there would have to be a special case for jQuery promises (or any that have a `state()` method) and other thenables.

I might be mistaken but I think it is possible to maintain an synchronous behavior of promises with `when` / `done`, etc..., while being fully Promises/A+ compliant.

Otherwise, I'll have to re-write `Parsley`, which combines synchronous and sometimes asynchronous validations very elegantly, and can see at the end if the complex combination of promises was comprised solely of synchronous checks or if there was an asynchronous one somewhere in there. It's a genuine use case that has no good solution with purely asynch promises :cry:

**Comment by gibson042** [comment]
> Mozilla's doc states that "Promise.resolve(value) method returns a Promise object that is resolved with the given value. If the value is a thenable [...]; otherwise the returned promise will be fulfilled with the value."

> I would expect the equivalent to have the corresponding state 'resolved', not 'pending'.

[ES6 Promises](http://www.ecma-international.org/ecma-262/6.0/#sec-promise-objects) are three-state pending/fulfilled/rejected, with _additional_ concepts of "settled" (not pending) and "resolved" (either settled, or locked in to match the state of another thenable, and in either case ignoring resolve/reject attempts). So not only is "resolved" not a state, but it is _completely_ independent of state—"A resolved promise may be pending, fulfilled or rejected."
jQuery Deferreds are three-state pending/resolved/rejected. It is unfortunate that our "resolved" state maps to ES6 "fulfilled", but to be fair, we were first.
At any rate, both `Promise.resolve(42)` and (as of 356a3bccb0e7468a2c8ce7d8c9c6cd0c5d436b8b) `jQuery.when(42)` are in state "pending" until the execution context stack fully clears.

> It's only with `jQuery`'s `done` and `state` methods that we can talk about synchronous behavior. Using `then` on jQuery would still be asynch and in line with other implementations even in the ABCD case.

Right, and that's what we did with Deferred. But Deferred doesn't unwrap thenables, and `jQuery.when( value )` does, which was requested both to work around the Deferred behavior and to better track `Promise.resolve`.

> I'm not sure I made myself clear, but `Promise.resolve(somePromise)` returns the *exact same object* `somePromise`, not a new promise. `$.when(aPromise)` returns a *different* object, that doesn't even have the same `state`.

`Promise.resolve(PromiseFromOtherRealm)` still returns a new object, but I get your point. Now that `jQuery.when(v)` is suitably distinguished from `jQuery.when(v1, …)`, we could reconsider the "jQuery.when always returns a new Deferred" decision from gh-2018. I'm not sure the functionality is worth the size, though.

> Indeed, there would have to be a special case for jQuery promises (or any that have a `state()` method) and other thenables.

`jQuery.when` tries to guarantee Promises/A+ compatibility, which it can't do if it blindly passes through e.g. a jQuery 1.x Deferred. So we could only pass through _our_ Deferreds, which would require some way of identifying them (e.g., changing from a factory to an inheritance model).

> Otherwise, I'll have to re-write `Parsley`, which combines synchronous and sometimes asynchronous validations very elegantly, and can see at the end if the complex combination of promises was comprised solely of synchronous checks or if there was an asynchronous one somewhere in there. It's a genuine use case that has no elegant solution with purely asynch promises :cry:

Can you describe exactly what you need from us? `jQuery.when` may very well be the wrong tool for the job, especially given that Deferred `done` is still sometimes synchronous.

**Comment by markelog** [comment]
> Promise.resolve(PromiseFromOtherRealm) still returns a new object, but I get your point.

mmm, i think we are went to far with analogies. Which is confusing, from which i would propose to eliminate.

For me, `jQuery.when(42)` looks analogues to `Promise.all([42])`, which is also in `pending` state and resolved only at the next tick.

But `Promise.all([42, 43])` will also be in `pending` and `jQuery.when(42,43)` won't be, so i would propose for it to be the only analogie - `Promise.all == jQuery.when`. 

How we map states es6 is looks clear to me, it seems we have just different name for the same thing.

> Can you describe exactly what you need from us? 

It is good we have this discussion though, that is why we did all those alphas and betas. But right now i think, we need to make this consistent.

**Comment by dmethvin** [comment]
> But right now i think, we need to make this consistent.

The word "consistent" will get us nowhere because there are so many ways to be consistent with a method like this that does disjoint things. Even @marcandre mentions there would need to be a special case for jQuery Deferred in order to get the output behavior he wants, which is another way to say the input isn't handled consistently. :smiling_imp: 

> In closing, I'd ask naively: why not have two different methods?

As @gibson042 said above, we completely agree and do plan to have two methods to separate out the two behaviors `$.when` is trying to accomplish. He's already [prepared the code](https://github.com/jquery/jquery/commit/356a3bccb0e7468a2c8ce7d8c9c6cd0c5d436b8b) for that. Your input is good to help us decide what additional methods we need in 3.1. For now it would be glorious to ship 3.0 because that is how we'll get feedback about what we missed. 

> Can you describe exactly what you need from us?

That is the right question. "I'll have to re-write" is a bit hyperbolic. The solution to avoid a rewrite doesn't even need to be inside the jQuery file, it could be a small plugin that retains whatever you need of the old behavior.  For better or worse, async ES6 `Promise` behavior is what developers are going to expect in the future and we are just going to confuse people by doing things differently, even if some people think it is better.

**Comment by marcandre** [comment]
I really appreciate your consideration and replies, even though I am not a known jQuery contributor :smile:

> At any rate, both Promise.resolve(42) and (as of 356a3bc) jQuery.when(42) are in state "pending" until the execution context stack fully clears.

I believe this is incorrect. `Promise.resolve(42)` has state 'fulfilled' (which is what I meant when I used jQuery's 'resolved', sorry about the confusion).

`Promise.all([42])` has state 'pending' until next tick. This has no consequence for browsers as there's only an asynch API anyways, so `foo` would execute at exactly the same moment if written as `Promise.resolve(42).then(foo)` or `Promise.all([42, 42]).then(foo)`.

`jQuery` promise has a synchronous API with `done`, so it does matter. My wish would be that the execution time of `$.when(42).then(foo)` or `$.when(42, 42).then(foo)` is the same as with native promises (i.e. at next tick), but that using `done` instead of `then` would also lead to the same execution time and that it executes synchronously (i.e. immediately)

> Can you describe exactly what you need from us?

tldnr: In short, the *only code I'd like to change* is so that `$.when(single)` returns a promise that behaves exactly like `$.when(single, 42)` does currently. I'd gladly provide a PR. Ok, not quite true, I'd like `$.when().state` to return `"resolved"` too.

longer:
I think we all desire an API compatible with native `Promise.all` and `promise.then`. Looks like 3.1 will introduce an equivalent to `Promise.resolve`. Good.

My wish is additionally that jQuery maintains a synchronous API that would fulfill jQuery promises (not other thenables) as soon as possible, so that any combination of `when` with already fulfilled jQuery promises yield an already fullfilled promise. Note that this is *exactly* what you are doing currently, *except* for `$.when(single)` and `$.when()`.

I believe that the special treatment of jQuery promises by `$.when` has no implication for the asynch API (since `then` on fulfilled promises still executes next tick). Moreover the special treatment of jQuery promises by the jQuery API is similar to the special treatment of native promises by `Promise.resolve`. Note that there's *already* a special treatment in `$.when(many, arguments)`:

```
p = $.Deferred()
p.resolve()
$.when(p,p).done(function() { console.log("A")})
n = Promise.resolve(42);
$.when(n,n).done(function() { console.log("B")})
console.log("C")
```
This outputs ACB (which I like very much and hope does not change). The output when using `then` is consistent (CAB, which we all want).

**Comment by gibson042** [comment]
> For me, `jQuery.when(42)` looks analogues to `Promise.all([42])`, which is also in `pending` state and resolved only at the next tick… i would propose for it to be the only analogie - `Promise.all == jQuery.when`. 

Definitively _not_. `jQuery.when(42)` resolves to a scalar (like `Promise.resolve`), but `Promise.all` can resolve only to an array.

**Comment by gibson042** [comment]
Thanks very much for the elaboration, @marcandre.

> `jQuery` promise has a synchronous API with `done`, so it does matter. My wish would be that the execution time of `$.when(42).then(foo)` or `$.when(42, 42).then(foo)` is the same as with native promises (i.e. at next tick), but that using `done` instead of `then` would also lead to the same execution time and that it executes synchronously (i.e. immediately)
> …
> My wish is additionally that jQuery maintains a synchronous API that would fulfill jQuery promises (not other thenables) as soon as possible, so that any combination of `when` with already fulfilled jQuery promises yield an already fullfilled promise. Note that this is *exactly* what you are doing currently, *except* for `$.when(single)` and `$.when()`.

This is possible, and would match our Deferred migration treatment. Any thoughts, @dmethvin or @markelog?

**Comment by markelog** [comment]
@gibson042

I wrote "analogie" here, not "equvivalent", even used `==` not `===`. Different output is fine to me.

> This is possible, and would match our Deferred migration treatment. Any thoughts, @dmethvin or @markelog?

What would be the purpose? To make it faster? I'm with this guy on this - https://plus.google.com/+PaulIrish/posts/ZBP64F2GEY4

Especially in the browser environment, besides there is really easy to make a plugin for it. 

Not really oppose this though, would i want is to make this as close to the standard as possible and make the output predictable.

**Comment by markelog** [comment]
Did a quick search in bluebird/q/when can't find anything like this, so after some thought i would be in favour of deprecating `done` then to make it more friendly to `jQuery.when`.

Also from reading the spec promises are 

> A promise represents the eventual result of an asynchronous operation.

So if `when` returns promise it should be asynchronous, seems if we would do this, we would be moving (although open for interpretation) from the spec, which doesn't seems ideal

**Comment by marcandre** [comment]
> seems if we would do this, we would be moving (although open for interpretation) from the spec, which doesn't seems ideal

The spec states that `then(callback)` should always run the callback asynchronously. That would not change. We all agree to be compatible with the spec.

**Comment by gibson042** [comment]
@markelog: Those are both missing the point somewhat—`.then` is spec-compliant, `.done` is backwards compatible. With Deferred, we pretty much maintained `.done` synchronicity. Likewise with multi-argument `jQuery.when`. The question now is about no-argument and single-argument `jQuery.when`.

**Comment by markelog** [comment]
> We all agree to be compatible with the spec.

It would be compatible with the spec, but would be moving away from its spirit and practise of other implementations. 

The point of the changing behaviour of the Deferreds was to do the opposite. I would treat `done` method as legacy, such legacy that could be deprecated in the future and eventually replaced by the native promises. By if we would move away from the spec, it means it would harder and harder to do it. 

> we pretty much maintained .done synchronicity

it could be synchronous still, could be asynchronous, in the case of `when` it could be just always be async.

Since promises are all about deferred computations, at least they are perceived that way - the whole point of making `then` to always be async was to give users predictable behaviour of async and *possible* async operations, that which we were oppose from the begging - we were wrong, community adopted and liked different behaviour.

Now it seems there is the same story, so yeah, we could do that and it would be back-compat, i just don't see the point.

**Comment by dmethvin** [comment]
> This is possible, and would match our Deferred migration treatment. Any thoughts, @dmethvin or @markelog?

I suppose we could treat `$.when` like the confusing API that it is, deprecate it as of 3.0, and seal it into the cellar with a cask of Amontillado. In that case we could value backcompat over consistency and spec compliance. That sets the stage to remove it in 4.0 and support it via Migrate.

**Comment by markelog** [comment]
@dmethvin that seems very radical solution, there is plenty legitimated cases where `jQuery.when` is very appropriated, personally, i don't find it confusing just like i don't perceive `Promise.all` confusing.

Since proposition of this ticket to adhere to `Promise.all` logic and every other implementations out there.

Right now it is in transitive state, which is understandable since it was pretty big rewrite of internal logic, so as is it is confusing but resolving this ticket as intendented would solve consistency issue without introducing any new quirks.

**Comment by marcandre** [comment]
> I would treat `done` method as legacy, such legacy that could be deprecated in the future and eventually replaced by the native promises.

That's one possible avenue. The other solution is to embrace having a synchronous API (`done`, `pipe`) and an async one (`then`).

> we were wrong, community adopted and liked different behaviour

I found the [original thread](https://github.com/promises-aplus/promises-spec/issues/4
) where this was decided. Everyone agrees on having the behavior well defined, but the community was/is quite split on if it should be synch or asynch. The only reason for the asynch behavior choice seems to be less error prone (although nobody pointed an actual real life error caused by synch evaluation). Moreover, there wasn't a good usecase presented for the synchronous API at the time, sadly (where was I? 😆 ), so the cost of preventing some legitimate uses was underestimated imo.

**Comment by dmethvin** [comment]
I don't think we can or want to make `$.when` the same as `Promise.all` because that will break @marcandre's use case just as much as it is right now (just a different way), plus it will break many other usages even more.

At this point `$.when` is synchronous (sometimes not), resolves with an array of values (sometimes not), is like `Promise.resolve` (sometimes not), is like `Promise.all` (sometimes not), takes multiple ticks to resolve its inputs (sometimes not), and so on. 

To @gibson042's question explicitly, the [documentation](http://api.jquery.com/jQuery.when/) already specifies a lot of these special cases so we can save work in the docs and compatibility issues by making it continue to work that way for all the special cases. We just need to expand the explanations for what happens if it's a single thenable or a mix of multiple thenables and deferreds. Luckily if you're looking at multiple Deferreds it seems much less likely that you depended on them bing 

> That's one possible avenue. The other solution is to embrace having a synchronous API (done, pipe) and an async one (then).

That is the case in jQuery today although it's more like resigned coexistence than an embrace. :smile_cat: 

> Everyone agrees on having the behavior well defined, but the community was/is quite split on if it should be synch or asynch. 

Perhaps there is a still a good place to litigate the point but I wouldn't say it is here. I would still be in favor of deprecating `$.when` since it has so many crazy exceptions that make it hard to use consistently.

**Comment by timmywil** [comment]
Will be closed by #3105
